azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
         return agent_family_manifests[0]
 
     @staticmethod
-    def __get_requested_version(agent_family):
+    def _get_version_from_gs(agent_family):
         """
-        Get the requested version from agent family
-        Returns: Requested version if supported and available in the GS
-                 None if requested version missing or GA versioning not enabled
+        Get the version from agent family
+        Returns: version if supported and available in the GS
+                 None if version is missing
         """
-        if conf.get_enable_ga_versioning() and agent_family.is_requested_version_specified:
-            if agent_family.requested_version is not None:
-                return FlexibleVersion(agent_family.requested_version)
+        if agent_family.version is not None:
+            return FlexibleVersion(agent_family.version)
         return None
 
-    @staticmethod
-    def __get_largest_version(agent_manifest):
-        largest_version = FlexibleVersion("0.0.0.0")
-        for pkg in agent_manifest.pkg_list.versions:
-            pkg_version = FlexibleVersion(pkg.version)
-            if pkg_version > largest_version:
-                largest_version = pkg_version
-        return largest_version
+    def run(self, goal_state):
+
+        try:
+            # Ignore new agents if update is disabled. The latter flag only used in e2e tests.
+            if not self._autoupdate_enabled or not conf.get_download_new_agents():
+                return
+
+            agent_family = self._get_agent_family_manifest(goal_state)
+            version = self._get_version_from_gs(agent_family)
+            gs_id = goal_state.extensions_goal_state.id
+            self._update_rsm_version_state_if_changed(goal_state.extensions_goal_state.created_on_timestamp, agent_family)
+            # if version is specified and vm is enabled for rsm upgrades, use rsm update path, else sef-update
+            if version is None and self._is_vm_enabled_for_rsm_upgrades and self._is_version_from_rsm:
```

Comment:
this check should be done when we are picking up those properties from the goal state